gems/pending/openstack/openstack_handle/handled_list.rb

```diff
@@ -0,0 +1,118 @@
+      []
+    end
+
+    private
+
+    def get_marker_based_paginated_list(collection_type, options, method)
+      all_servers = call_list_method(collection_type, options, method)
+      last_server = all_servers.try(:last)
+
+      while marker(last_server, collection_type)
+        servers_on_page = call_list_method(collection_type,
+                                           options,
+                                           method,
+                                           :marker => marker(last_server, collection_type),
+                                           :limit  => default_pagination_limit)
+        last_server = servers_on_page.try(:last)
+        break if pagination_break_condition(all_servers, servers_on_page, collection_type)
+        all_servers.concat(servers_on_page) unless servers_on_page.blank?
+      end
+
+      all_servers
+    end
+
+    def get_page_number_based_paginated_list(_collection_type, _options, _method)
+      # TBD, used e.g. by keystone v3
+    end
+
+    def call_list_method(collection_type, options, method, paginate_options = {:limit => default_pagination_limit})
+      public_send(collection_type).public_send(method, options.merge(paginate_options))
+    end
+
+    def marker(last_server, collection_type)
+      last_server.send(identity_method(collection_type)) if last_server && identity_method(collection_type)
+    end
+
+    def pagination_type(_collection_type)
+      # Using method, so we can e.g set pagination type per method name, e.g. when some collection doesn't support
+      # pagination, like Heat resources, but others do
+      # Allowed values :marker, :page_number, :none
+      :marker
+    end
+
+    def multi_tenancy_type(_collection_type)
+      # Using method, so we can e.g set multi_tenancy_type type per method name, e.g. when attribute all_tenants is
+      # broken on some collections, so it's better to rather sent API request per tenant
+      # Allowed values :all_tenants_loop, :page_number, :none
+      :all_tenants_loop
+    end
+
+    def identity_method(_collection_type)
+      # Identity method for the collection type
+      :id
+    end
+
+    def default_pagination_limit
+      1000
+    end
+
+    def pagination_break_condition(all_servers, servers_on_page, collection_type)
+      # Test if the whole set of records isn't already present, if it is, break the pagination.
+      # E.g. Neutron can have disabled pagination like this, then it just returns the same result and pagination
+      # would loop forever.
+      # There can be indication if there is another page, e.g. neutron should have next link, but fog collection
+      # doesn't support any attributes, so we will need to add it there. That will be much more effective than this
+      # solution. Then services supporting this can override this method.
+      (servers_on_page.collect(&identity_method(collection_type)) -
+       all_servers.collect(&identity_method(collection_type))).count <= 0
```

Comment:
Possibly, but then it will be order dependent. So I rather chose slower but more reliable way. But I guess I can trust the order will be always the same. :-)